tasks/generateSettings.js

```diff
@@ -0,0 +1,278 @@
+
+const LEAGACY_SETTINGS_PROJECT_DIR = 'source/common/res/features';
+const NEW_SETTINGS_PROJECT_DIR = 'sauce/features';
+const ALL_SETTINGS_OUTPUT = 'source/common/res/features/allSettings.js';
+const REQUIRED_SETTINGS = ['name', 'type', 'default', 'section', 'title'];
+
+let previousSettings;
+
+function run(callback) {
+  previousSettings = new Set();
+  Promise.all([
+    gatherLegacySettings(),
+    gatherNewSettings()
+  ]).then(values => {
+    let validatedSettings = [];
+    let settingsConcatenated = values[0].concat(values[1]);
+    settingsConcatenated.forEach(setting => {
+      if (Array.isArray(setting.setting)) {
+        setting.setting.forEach(subSetting => {
+          let validatedSetting = validateSetting({
+            setting: subSetting,
+            file: setting.file
+          });
+
+          if (validatedSetting.hidden !== true) {
+            validatedSettings.push(validatedSetting);
+          }
+        });
+      } else {
+        let validatedSetting = validateSetting(setting);
+
+        if (validatedSetting.hidden !== true) {
+          validatedSettings.push(validatedSetting);
+        }
+      }
+    });
+
+    let allSetingsFile = generateAllSettingsFile(validatedSettings);
+    fs.writeFile(ALL_SETTINGS_OUTPUT, allSetingsFile, callback);
+  }, reason => {
+    callback(reason);
+  }).catch(exception => {
+    callback(exception.stack);
+  });
+}
+
+function gatherLegacySettings() {
+  return new Promise((resolve, reject) => {
+    glob(`${LEAGACY_SETTINGS_PROJECT_DIR}/**/settings.js*`, (error, files) => {
+      if (error) return reject(error);
+
+      let legacySettingsPromises = [];
+
+      files.forEach(file => {
+        legacySettingsPromises.push(new Promise((resolve, reject) => {
+          let setting;
+          const filePath = `${__dirname}/../${file}`;
+          try {
+            setting = require(filePath); // eslint-disable-line global-require
+          } catch (e) {
+            fs.readFile(filePath, 'utf8', (error, data) => {
+              if (error) return reject(error);
+              setting = JSON.parse(data);
+            });
+          }
+
+          resolve({ file, setting, legacy: true });
+        }));
+      });
+
+      Promise.all(legacySettingsPromises).then(resolve, reject);
+    });
+  });
+}
+
+function gatherNewSettings() {
+  return new Promise((resolve, reject) => {
+    glob(`${NEW_SETTINGS_PROJECT_DIR}/**/settings.js`, (error, files) => {
+      if (error) return reject(error);
+
+      resolve(files.map(file => {
+        const setting = require(`${__dirname}/../${file}`); // eslint-disable-line global-require
+        return { file, setting };
+      }));
+    });
+  });
+}
+
+function validateSetting(settingObj) {
+  const featureSettings = settingObj.setting;
+  const settingFilename = settingObj.file;
+
+  REQUIRED_SETTINGS.forEach(requiredSetting => {
+    if (typeof featureSettings[requiredSetting] === 'undefined') {
```

Comment:
Should `null` be considered a violation of the requirement as well? I'm completely ok if it gets treated like false, just would be odd to find `null` settings. Not my strongest held opinion ever. 😄